compiler/rustc_passes/src/stability.rs

```diff
@@ -772,13 +772,31 @@ impl<'tcx> Visitor<'tcx> for Checker<'tcx> {
                 let ty = self.tcx.type_of(item.def_id);
                 let ty::Adt(adt_def, substs) = ty.kind() else { bug!() };
 
-                // Non-`Copy` fields are unstable, except for `ManuallyDrop`.
+                #[allow(rustc::usage_of_qualified_ty)] // `Ty` is the wrong type here, we really want `ty::Ty`.
+                fn allowed_union_field<'tcx>(
+                    tcx: TyCtxt<'tcx>,
+                    param_env: ty::ParamEnv<'tcx>,
+                    ty: ty::Ty<'tcx>,
+                ) -> bool {
+                    // We don't just accept all !needs_drop fields, due to semver concerns.
+                    match ty.kind() {
+                        ty::TyKind::Ref(..) => true, // references never drop (even mutable refs, which are non-Copy and hence fail the later check)
```

Comment:
I noticed that `is_trivially_pure_clone_copy` is, for some reason, careful around references and raw pointers:

https://github.com/rust-lang/rust/blob/90db033955eaa000f8eea5691ea4297687a4dbef/compiler/rustc_middle/src/ty/sty.rs#L2340

Not sure if that should be a concern here.

Reply:
at least mutable references can't be cloned or copied, so it makes sense.

Reply:
Hm, the comment on that calls out custom metadata ("Thin pointers & thin shared references are pure-clone-copy, but for anything with custom metadata it might be more complicated."), which I think makes sense for Copy/Clone -- we might eventually support having metadata be !Copy, I guess. But here we really care about !Drop, not !Copy, and I can't ever see `&T`/`&mut T`/`*const`/`*mut` gaining any kind of drop impl, so I think we're OK here.